OA-PASS/pass-docker
# Adds Deposit Services

Adds a `deposit` container with the 0.0.3 release of Deposit Services.
- Bumps fcrepo container tag
    - Includes updated `logback.xml` (fixed width fields, thread name)
    - Updated `jms-addon` (improved logging)
    - Updates JMS selectors to use with the latest PASS namespace
    - Includes updated pass-user-service version 0.0.3 with improved logging and docker-machine compatibility
- Adds `oapass/deposit-services:0.0.3-2.2` image
    - Includes/updates deposit services runtime env vars to `.env`
- Removes defunct `nihms-submission` image
- Bumps the `proxy` image tag
    - Allows the DSpace `xmlui` to be viewed through the proxy
- Updates the docker-compose configuration to allow DSpace to present publicly-resolvable URLs

To test:
- Check out this PR
- Clear everything out of docker; `docker-compose down -v` should be sufficient
- (Re)build the `fcrepo`, `deposit`, and `proxy` containers:
    - `docker-compose build --no-cache fcrepo deposit proxy`
    - if there are problems testing this PR, such as connectivity to elastic search (you'll see red requests in your javascript/dev tools console when requests are made to ES), you may need to rebuild the `sp` container as well (not sure why...)
- Edit .env to suit
- Start everything up `docker-compose -d`
- Tail the logs from the `deposit` container (`docker logs -f deposit`)
    - wait for a message: 
> 16:44:20.114 [      main] INFO  [ ListenerRunner] - Fedora repository is up at 'http://192.168.99.100:8080/fcrepo/rest/'

- Visit `https://pass` in browser, and either: run the dev tools console (`alt-cmd`-`i`), or use your browser's incognito mode (to disable the browser caching)
    - log in as `faculty1`
- Next, perform a submission by going to the "Grants" tab, and clicking "New Submission" for the project named `Digital Surrogates Of Medieval Manuscripts: Le Roman De La Rose Phase Ii`
    - I've been using DOI `10.1021/acsmedchemlett.7b00397`
    - Accept the JScholarship deposit agreement
    - Attach at least one file
    - Feel free to empty the scrollback buffer of the terminal tailing the `deposit` logs
    - click submit
- Watch deposit services logs, wait for messages like (the `Deposit-Worker` or `DSpace Archive Writer` threads are the interesting ones if you want to `grep`):
> 17:00:22.303 [Deposit-Worker-0] DEBUG [    DepositTask] - >>>> Deposit http://192.168.99.100:8080/fcrepo/rest/deposits/a9/90/62/ad/a99062ad-52b6-472e-bcc1-bb32b24989a2 was accepted.
> 17:00:22.608 [Deposit-Worker-0] DEBUG [    DepositTask] - >>>> Created repository copy for http://192.168.99.100:8080/fcrepo/rest/deposits/a9/90/62/ad/a99062ad-52b6-472e-bcc1-bb32b24989a2: http://192.168.99.100:8080/fcrepo/rest/repositoryCopies/b4/38/2f/a9/b4382fa9-99c4-4935-91b5-98910147e19b
> 17:00:22.610 [Deposit-Worker-0] DEBUG [    DepositTask] - >>>> Updating deposit http://192.168.99.100:8080/fcrepo/rest/deposits/a9/90/62/ad/a99062ad-52b6-472e-bcc1-bb32b24989a2
- You should see messages like the above in the `DepositTask` thread:
    - `Deposit <uri> was accepted.`
    - `Created repository copy for ...`
- Ending with:
    - `Updating deposit <uri>`
- Navigate to the deposit, in the above example http://192.168.99.100:8080/fcrepo/rest/deposits/a9/90/62/ad/a99062ad-52b6-472e-bcc1-bb32b24989a2
    - Verify its status is 'accepted'
- Navigate to the submission
    - Verify the aggregated deposit status is 'accepted'
- Navigate back to the Deposit, then to its repository copy
    - Verify its status is 'complete'
- Copy and paste the `externalId` of the repository copy into your browser
    - you should be taken to DSpace and see your deposit
- The DSpace "Full item record" should have sane metadata
- Ought to be able to download the file(s) of the submission, even if the file names are not correctly displaying (see known issues)
- Finally, back in the Ember UI you ought to be able to view the list of submissions, and while the table is lacking information, you should see a green dot with "Accepted" for its status.

Known issues:  
- Exceptions occur during startup of deposit services, and when handing the the submission.  Shouldn't affect anything.
- File names in the DSpace submission are not carried over from the names of the PASS `File` object

If you encounter issues, please send/attach the full output of `docker logs deposit`, and if you feel like it, the full output of `docker logs fcrepo`
